Automatically delete tags after branch is deleted

gvegayon opened the pull request:
This pull request introduces a new GitHub Actions workflow to delete tags from a container registry when a branch is deleted. The workflow includes steps for checking out the code, determining the image tag, logging into the container registry and Azure, and finally deleting the specified images.

Key changes:

* [`.github/workflows/delete-container-tag.yaml`](diffhunk://#diff-5d41f68a88061c49f4d2b5b87b46ffdc4621806ca0e57829d75f5129b627c5e4R1-R59): Added a new workflow named "Delete tag from container registry" which is triggered on branch deletion. It includes steps to checkout code, determine the image tag, log into the container registry and Azure, and delete the specified images from the registry.

zsusswein commented on .github/workflows/delete-container-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+name: Delete tag from container registry
+
+on:
+  delete:
```
I don't believe this is the correct trigger. It should be on PR close because the container build + push happens on PR open, not on push.

zsusswein commented on .github/workflows/delete-container-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+on:
+  delete:
+
+env:
+  REGISTRY: cfaprdbatchcr.azurecr.io
+  USERNAME: cfaprdbatchcr
+  IMAGE_NAME: cfa-epinow2-pipeline
```
Can you use existing repository secrets rather than env variables?

zsusswein commented on .github/workflows/delete-container-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+
+jobs:
+  delete-container:
+    if: github.event.ref_type == 'branch'
```
Likewise, should be PR close.

zsusswein commented on .github/workflows/delete-container-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+      - name : Checkout code
+        uses: actions/checkout@v4
+
+      - name: Figure out tag (either latest if it is main or the branch name)
+        id: image-tag
+        run: |
+          if [ "${{ github.event.ref }}" = "main" ]; then
+            echo "The image associated with the main branch cannot be deleted."
+            exit 1
+          else
+            echo "tag=${{ github.event.ref }}" >> $GITHUB_OUTPUT
+          fi
```
We only need to pull the branch because this will never run on main.

zsusswein commented on .github/workflows/delete-container-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+          # handbook.
+          creds: ${{ secrets.EDAV_STF_SERVICE_PRINCIPAL }}
+
+      - name: Deleting the image
+        run: |
```
Can you modify this to use an [Azure CLI action](https://github.com/marketplace/actions/azure-cli-action#workflow-to-execute-an-azure-cli-script-of-the-latest-azure-cli-version) instead of using the version on the runner?

zsusswein commented on .github/workflows/delete-container-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+
+      - name: Deleting the image
+        run: |
+          # Remove the image from the registry
+          az acr repository delete \
+            --yes \
+            --name ${{ env.REGISTRY }} \
+            --image ${{ env.IMAGE_NAME }}:${{ steps.image-tag.outputs.tag }}
+
+          # Remove the dependencies image from the registry
+          az acr repository delete \
+            --yes \
+            --name ${{ env.REGISTRY }} \
+            --image ${{ env.IMAGE_NAME }}:dependencies-${{ steps.image-tag.outputs.tag }}
```
Can you move this code into a shell script like I did in #166? This script should take any Github environment variables or secrets as arguments so we can run it locally and should live in `.github/scripts`

zsusswein reviewed (changes requested):
Thanks George. See my feedback below.

gvegayon commented on .github/scripts/cleanup-tags-on-pr-close.sh:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+#
+# Delete Batch Pools and associated jobs
```
```suggestion
# Delete container tags from Azure CR
```

natemcintosh reviewed (approved):
I think this looks pretty good. Have we been able to test it out yet?

Hold on merging till @zsusswein has also given it another look; he's a little more up to date all this.

zsusswein commented:
Is this ready for a re-review? I was waiting until I was requested.

gvegayon commented on .github/scripts/cleanup-tags-on-pr-close.sh:
```diff
@@ -0,0 +1,31 @@
+
+echo "Logging into Batch account"
+az batch account login \
+  --name "${ACCOUNT_NAME}" \
+  --resource-group "${RESOURCE_GROUP}"
```
Great catch, nate! Are these ${{ secrets.BATCH_ACCOUNT_NAME }} ${{ secrets.RESOURCE_GROUP }} ?

gvegayon replied:
> I think this looks pretty good. Have we been able to test it out yet?

For this, we implemented this in pyrenew-hew. We should be able to test it as soon as we delete the branch :).

zsusswein reviewed (changes requested):
George can you take another look at the script and then request a re-review when it's ready? I don't think this is correct just yet. 

For example, the script doesn't currently run because it requires 3 arguments (line 5) but needs 5 arguments. It also logs into Batch but only ever interacts with the container registry.

gvegayon commented:
OK @zsusswein and @natemcintosh, I've fixed the issue with the number of parameters. I also updated the news, fixing the conflicts. I tested the script on the VAP, and it works. Note that the deletion of the dependencies image is targeted to be used with the new action (pending review #167). #167 should be merged soon as it is a more robust approach for building and caching tags.

zsusswein commented:
Re this:
>  It also logs into Batch but only ever interacts with the container registry.

Can you explain why it's logging into Batch or delete that section?

gvegayon commented:
> Re this:
> 
> > It also logs into Batch but only ever interacts with the container registry.
> 
> Can you explain why it's logging into Batch or delete that section?

Yes, I was using that to log in to Azure, but I realized it is not needed as we already had the action prior to calling the script. I just modified both the bash script and workflow file to reflect this. It should be in a good shape now!

zsusswein reviewed (approved):
Thanks George! This is a nice improvement to the repo.
